core/src/bytecode/ast/compat.rs

```diff
@@ -0,0 +1,541 @@
+//! Compatibility with the current stable Nickel version.
+//!
+//! This module defines a trait for converting from the representation used in stable Nickel to the
+//! new AST representation of the bytecode compiler, and implements it for the types defined in
+//! [crate::bytecode::ast].
+
+use super::{primop::PrimOp, *};
+use crate::term;
+
+/// Convert from the mainline Nickel representation to the new AST representation. This trait is
+/// mostly `From` with an additional argument for the allocator.
+///
+/// # Parameters
+///
+/// - `'ast`: the lifetime of the AST nodes, tied to the allocator
+/// - `'a`: the lifetime of the reference to the mainline Nickel object, which doesn't need to be
+///   related to `'ast` (we will copy any required data into the allocator)
+/// - `T`: the type of the mainline Nickel object ([term::Term], [term::pattern::Pattern], etc.)
```

Comment:
Hmm, it's true that it sounds like `'a` should be a "free" lifetime. I think it's useful to keep both direction of the conversion around just in case; maybe we'll want to test both mainline and the bytecode vm for full runs and compare the result of each phase for example, I don't know. We'll get rid of both of them eventually anyway. So let me try to move the lifetime to `from_mainline`.

Reply:
Since it becomes the only lifetime parametrized at the level of the function (`'ast` is coming from `impl<'ast>`), we can just elide it entirely, which is even cleaner. Thanks!